apps/knowledge-base/src/providers/sdk-provider.tsx

```diff
@@ -0,0 +1,104 @@
+import * as React from 'react';
+import {
+  AppExtensionSDK,
+  SidebarExtensionSDK,
+} from 'contentful-ui-extensions-sdk';
+import { RateLimit } from 'async-sema';
+
+interface SdkContextProps {
+  getContentType?: (id: string) => Promise<Record<string, any>>;
+  updateContentType?: (
+    contentType: Record<string, any>
+  ) => Promise<Record<string, any>>;
+  createContentType?: (
+    contentType: Record<string, any>
+  ) => Promise<Record<string, any>>;
+  getEntry?: (id: string) => Promise<Record<string, any>>;
+  publishEntry?: (entry: Record<string, any>) => Promise<Record<string, any>>;
+  createEntry?: (
+    entryId: string,
+    entryData: Record<string, any>
+  ) => Promise<any>;
+  getAsset?: (id: string) => Promise<Record<string, any>>;
+  publishAsset?: (asset: Record<string, any>) => Promise<Record<string, any>>;
+  createAsset?: (asset: Record<string, any>) => Promise<Record<string, any>>;
+  processAsset?: (
+    asset: Record<string, any>,
+    locale: string
+  ) => Promise<Record<string, any>>;
+  prepareAsset?: (
+    asset: Record<string, any>,
+    locale: string
+  ) => Record<string, any>;
+  waitUntilAssetProcessed?: (asset: Record<string, any>, locale: string) => any;
+  instance?: AppExtensionSDK & SidebarExtensionSDK;
+}
+
+interface SdkProviderProps {
+  sdk: AppExtensionSDK & SidebarExtensionSDK;
+}
+
+const { createContext, useContext } = React;
+
+const SdkContext = createContext<SdkContextProps>({});
+
+const waitForRateLimit = RateLimit(8); // rate limit is 10/s, using 8 just to compensate any latency
```

Comment:
Rate limits should be handled just fine by the Web App. I think, if you've run into any issues, this could be the use of single-entity access endpoints (`getContentType`, `getEntry`...)

Reply:
I've received a lot of 429 from the API, not only from those but also when creating entries and mainly assets, one example is using the `waitUntilAssetProcessed` method, which hits the API constantly to know when the asset processing is done.